File: src/c07_methods/TemplateMethod.java
package c07_methods;

// Abstract class defining the template method
abstract class DocumentGenerator {
    // Template method
    public final void generateDocument() {
        collectData();
        formatContent();
        saveDocument();
    }

    protected abstract void collectData();
    protected abstract void formatContent();
    protected abstract void saveDocument();
}

// Concrete class implementing PDF document generation
class PDFGenerator extends DocumentGenerator {
    @Override
    protected void collectData() {
        System.out.println("Collecting data for PDF...");
    }

    @Override
    protected void formatContent() {
        System.out.println("Formatting content for PDF...");
    }

    @Override
    protected void saveDocument() {
        System.out.println("Saving as PDF file...");
    }
}

// Concrete class implementing Word document generation
class WordGenerator extends DocumentGenerator {
    @Override
    protected void collectData() {
        System.out.println("Collecting data for Word document...");
    }

    @Override
    protected void formatContent() {
        System.out.println("Formatting content for Word document...");
    }

    @Override
    protected void saveDocument() {
        System.out.println("Saving as Word file...");
    }
}

// Client code
public class TemplateMethod {
    public static void main(String[] args) {
        System.out.println("Generating PDF Document:");
        DocumentGenerator pdfGenerator = new PDFGenerator();
        pdfGenerator.generateDocument();

        System.out.println("\nGenerating Word Document:");
        DocumentGenerator wordGenerator = new WordGenerator();
        wordGenerator.generateDocument();
    }
}

File: src/c07_methods/matlab/TemplateMethod.m
function TemplateMethod()
    fprintf('Generating PDF Document:\n');
    pdfGenerator = PDFGenerator();
    pdfGenerator.generateDocument();

    fprintf('\nGenerating Word Document:\n');
    wordGenerator = WordGenerator();
    wordGenerator.generateDocument();
end

% Abstract class defining the template method
classdef DocumentGenerator < handle
    methods (Sealed)
        % Template method
        function generateDocument(obj)
            obj.collectData();
            obj.formatContent();
            obj.saveDocument();
        end
    end
    methods (Abstract, Access = protected)
        collectData(obj)
        formatContent(obj)
        saveDocument(obj)
    end
end

% Concrete class implementing PDF document generation
classdef PDFGenerator < DocumentGenerator
    methods (Access = protected)
        function collectData(obj)
            fprintf('Collecting data for PDF...\n');
        end
        function formatContent(obj)
            fprintf('Formatting content for PDF...\n');
        end
        function saveDocument(obj)
            fprintf('Saving as PDF file...\n');
        end
    end
end

% Concrete class implementing Word document generation
classdef WordGenerator < DocumentGenerator
    methods (Access = protected)
        function collectData(obj)
            fprintf('Collecting data for Word document...\n');
        end
        function formatContent(obj)
            fprintf('Formatting content for Word document...\n');
        end
        function saveDocument(obj)
            fprintf('Saving as Word file...\n');
        end
    end
end
